Bump Kotlin to 2.0.0

hfhbd opened the pull request:
Draft PR because the PR will be split into smaller ones.

I switched to atomicfu in coroutines test due some expect/actual/sourceset hierarchy errors because stately don't support android native, before realizing the error was caused by sourceset hierarchy, see https://youtrack.jetbrains.com/issue/KT-64483 for context. The switch is not needed, but does not hurt anyway in tests.

hfhbd commented:
Sample fails due to https://youtrack.jetbrains.com/issue/KT-68295/K2-Actual-typealias-as-generic-type-in-expect-class, workaround for this use-case, use kotlinx-datetime

AlecKazakova commented:
looks good to me if we wanna bump to 2.0 release

hfhbd commented on runtime/api/runtime.api:
@AlecKazakova what about this abi change? Should I revert it?

dellisd replied:
Why does this ABI change exist? (Was this a Kotlin 2.0 optimization?)

We probably shouldn't break the ABI while we're still on SQLDelight 2.x

AlecKazakova replied:
is this a breaking change though? I assume the synthetic function would still be invokable by anyone who compiled against earlier versions

AlecKazakova replied:
if there is a way to upgrade to 2.0 without this ABI change I think that would be preferable though

JakeWharton replied:
For multiple reasons it is breaking:
- Anyone implementing `QueryResult` without overriding `val value` will no longer be able to link to the default because it no longer exists.
- `AsyncValue` users who were calling the `value` property will no longer be able to link because the return type of the bytecode method changed.

hfhbd commented:
@AlecKazakova as one big PR?

AlecKazakova commented on runtime/src/commonMain/kotlin/app/cash/sqldelight/db/QueryResult.kt:
i see they come from actual changes, what in 2.0 required these changes?

hfhbd replied:
Actually, there was only the inline warning that will be an error with 2.1

AlecKazakova commented:
> @AlecKazakova as one big PR?

if all the changes are required for 2.0 then yea im fine with a bigger PR, it wasn't that tough to review

hfhbd commented:
@AlecKazakova the switch from stately to atomicFU isn't required for Kotlin 2.0.0, but we only use it in test code so I would keep it if its okay.
I changed the dependency because I got some error related to the hierarchically test source sets (missing some groups), so I got a wall of text regarding wrong dependencies etc., so I switched to atomicFU because it supports all Kotlin targets but in the end, the missing android native support of stately wasn't the problem.
